doc/website/advanced/best-practice-for-testing.md

```diff
@@ -0,0 +1,263 @@
+While the AAA pattern provides a sensible structure, [ZOMBIES](http://blog.wingman-sw.com/tdd-guided-by-zombies) help you to find sensible test cases.
+```
+Z = Zero
+O = One
+M = Many (or More complex)
+B = Boundary Behaviors
+I = Interface Definition
+E = Exercise Exceptional Behavior
+S = Simple Scenarios, Simple Solutions
+```
+
+This can be separated into **ZOM** and **BIE** with **S** bridging them together.
+**ZOM** are often simple tests, like _a vector with zero items is empty_ or _a vector with one item is not empty_ or _a vector with N items has a size of N_.
+The **BIE** part takes care of edge cases like _adding one to Number::max saturates_ or _division by zero returns error_.
+The latter overlaps with a case from the **ZOM** part, which illustrates that these are not always clearly separated.
+
+**Exercise Exceptional Behavior** means to not only test the happy path, but also the negative one.
+Basically to test with silly inputs and check for a graceful behavior like the previous example with division by 0.
+The linked blog post explains [negative testing](https://www.softwaretestinghelp.com/what-is-negative-testing/) in a more thorough way.
+
+The catchwords can be used to draw simple scenarios which nicely fits to the AAA pattern. A non-exhaustive list of these scenarios are
+- overflow
+- underflow
+- wrap around
+- empty
+- full
+- out of bounds
+- timeouts
+- copy
+    - are the objects equal
+    - is the copy origin unchanged
+    - etc.
+- move
+    - is the move destination object cleaning up its resources
+    - is the move origin object in a defined but unspecified state
+    - etc.
+- etc.
+
+Following [Hyrum's Law](https://www.hyrumslaw.com/) loosely, given enough users, one will find ways to use the software in a way it was never imagined.
+Therefore, never underestimate the creativity of brilliancy/stupidity.
+
+In some cases it might be necessary to instantiate an object on the heap. While that's not allowed in production code, it is fine in test code.
+To avoid manual memory management with new/delete, smart pointer shall be used if possible.
+As a reminder, if a method takes a pointer to an object, this object can be instantiated on the stack and the address of this object can be passed to the method.
+A good reason to use the heap are large objects which might cause a stack overflow.
+Some operating system have a rather small stack of only a few kB, so this limit might be closer one might think.
+
+Sometimes it is necessary to access a pointer returned by a method.
+Using an `EXPECT_*` to check for a `nullptr` is not sufficient if the pointer is dereferenced afterwards.
+It might crash the test application if it indeed is a `nullptr`.
+The implementation of the test object might be broken and only after the successful test run it can be assumed to be correct.
+To prevent a crash, the `nullptr` check has to be done with an `ASSERT_*`. This gracefully aborts the current test case and continues with the next one.
```

Comment:
In general ASSERT is to be used if te test cannot continue if the condition is not true (e.g. nullptr, empty optional, non-existing element in an array and similar cases)